compiler/rustc_lint/src/noop_method_call.rs

```diff
@@ -0,0 +1,86 @@
+    /// struct Foo;
+    /// let foo = &Foo;
+    /// let clone: &Foo = foo.clone();
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Some method calls are noops meaning that they do nothing. Usually such methods
+    /// are the result of blanket implementations that happen to create some method invocations
+    /// that end up not doing anything. For instance, `Clone` is implemented on all `&T`, but
+    /// calling `clone` on a `&T` where `T` does not implement clone, actually doesn't do anything
+    /// as references are copy. This lint detects these calls and warns the user about them.
+    pub NOOP_METHOD_CALL,
+    Warn,
+    "detects the use of well-known noop methods"
+}
+
+declare_lint_pass!(NoopMethodCall => [NOOP_METHOD_CALL]);
+
+impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        // We only care about method calls
+        if let ExprKind::MethodCall(..) = expr.kind {
+            // Get the `DefId` only when dealing with an `AssocFn`
+            if let Some((DefKind::AssocFn, did)) =
+                cx.typeck_results().type_dependent_def(expr.hir_id)
+            {
+                // Check that we're dealing with a trait method
+                if let Some(trait_id) = cx.tcx.trait_of_item(did) {
+                    // Check we're dealing with one of the traits we care about
+                    if ![sym::Clone, sym::Deref, sym::Borrow]
+                        .iter()
+                        .any(|s| cx.tcx.is_diagnostic_item(*s, trait_id))
+                    {
+                        return;
+                    }
+
+                    let substs = cx.typeck_results().node_substs(expr.hir_id);
+                    // We can't resolve on types that recursively require monomorphization,
+                    // so check that we don't need to perfom substitution
+                    if !substs.needs_subst() {
+                        let param_env = cx.tcx.param_env(trait_id);
+                        // Resolve the trait method instance
+                        if let Ok(Some(i)) = ty::Instance::resolve(cx.tcx, param_env, did, substs) {
+                            // Check that it implements the noop diagnostic
+                            if [
+                                sym::noop_method_borrow,
+                                sym::noop_method_clone,
+                                sym::noop_method_deref,
+                            ]
+                            .iter()
+                            .any(|s| cx.tcx.is_diagnostic_item(*s, i.def_id()))
+                            {
+                                let span = expr.span;
+
+                                cx.struct_span_lint(NOOP_METHOD_CALL, span, |lint| {
+                                    let message = "call to noop method";
+                                    lint.build(&message).emit()
```

Comment:
I think `MachineApplicable` is too strong - there are some corner cases where the method call needs to be replaced with a dereference.

Reply:
We can change it to `MaybeIncorrect`, not too bothered either way :)

Reply:
So I added this suggestion, and I wasn't satisfied with the results. Some method calls aren't _technically_ noops in that they can for instance turn a `&&` doouble reference into a `&` single reference. Applying this suggestion naively can lead to code that doesn't compile. For instance:

```rust
use std::borrow::Borrow;
struct Foo<T>(T);
let a = &&Foo(1u32);
let borrowed: &Foo<u32> = a.borrow();
```

This code compiles and produces a warning because there's no need to call `borrow` here. However, removing the call to `borrow` will break this code the type annotation requires that `a` be of type `&Foo<u32>` not `&&Foo<u32>`.

In any case if we do decide to add the suugestion back it's as simple as adding:
```rust
.span_suggestion_verbose(
  // remove one from method call span for the `.`
  expr_span.with_lo(method_call_span.lo() - BytePos(1)),
  "remove the unnecessary method call",
  String::new(),
  Applicability::MaybeIncorrect,
)
 ```

Reply:
Side note, my preference when creating new spans is to take to existing spans and operate on them. `""     .    to_string()` is valid code and doing operations with `BytePos` is a way of creating invalid spans.

Reply:
I couldn't figure out how to get the span to encompass the `.` of the method call without doing `BytePos` arithmetic. Is there a better way?